File: card_recognizer.py
"""
Card recognition module for MTG card estimator.
Identifies MTG cards using the Scryfall API.
"""
import requests
import numpy as np
from typing import Optional, Dict, Any
import base64
import io
from PIL import Image


class CardRecognizer:
    """Recognizes MTG cards using Scryfall API."""
    
    def __init__(self):
        """Initialize the card recognizer."""
        self.scryfall_api_base = "https://api.scryfall.com"
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'MTG-Estimator/1.0'
        })
    
    def search_card_by_name(self, card_name: str) -> Optional[Dict[str, Any]]:
        """
        Search for a card by name using Scryfall API.
        
        Args:
            card_name: Name of the card to search for
            
        Returns:
            Card data from Scryfall or None if not found
        """
        try:
            url = f"{self.scryfall_api_base}/cards/named"
            params = {'fuzzy': card_name}
            response = self.session.get(url, params=params)
            
            if response.status_code == 200:
                return response.json()
            else:
                return None
        except Exception as e:
            print(f"Error searching for card: {e}")
            return None
    
    def get_card_info(self, scryfall_id: str) -> Optional[Dict[str, Any]]:
        """
        Get detailed card information by Scryfall ID.
        
        Args:
            scryfall_id: Scryfall UUID for the card
            
        Returns:
            Card data from Scryfall or None if not found
        """
        try:
            url = f"{self.scryfall_api_base}/cards/{scryfall_id}"
            response = self.session.get(url)
            
            if response.status_code == 200:
                return response.json()
            else:
                return None
        except Exception as e:
            print(f"Error getting card info: {e}")
            return None
    
    def recognize_card(self, card_image: np.ndarray) -> Optional[Dict[str, Any]]:
        """
        Recognize a card from an image.
        Note: This is a simplified implementation. In a production system,
        you would use OCR or image recognition to extract the card name.
        
        Args:
            card_image: OpenCV image of the card
            
        Returns:
            Card data or None if not recognized
        """
        # For now, this is a placeholder that would need OCR integration
        # In a real implementation, you would:
        # 1. Use OCR (like Tesseract) to extract text from the card name area
        # 2. Search for the card using the extracted name
        # 3. Possibly use image matching for the card art
        
        # Placeholder: Return None as we need OCR for actual recognition
        return None
    
    def search_cards(self, query: str, page: int = 1) -> Dict[str, Any]:
        """
        Search for cards using Scryfall search syntax.
        
        Args:
            query: Search query
            page: Page number for pagination
            
        Returns:
            Search results from Scryfall
        """
        try:
            url = f"{self.scryfall_api_base}/cards/search"
            params = {'q': query, 'page': page}
            response = self.session.get(url, params=params)
            
            if response.status_code == 200:
                return response.json()
            else:
                return {'object': 'error', 'details': 'Search failed'}
        except Exception as e:
            print(f"Error searching cards: {e}")
            return {'object': 'error', 'details': str(e)}

File: app/services/card_recognizer.rb
# frozen_string_literal: true

require 'httparty'

class CardRecognizer
  include HTTParty
  base_uri 'https://api.scryfall.com'

  def initialize
    self.class.headers('User-Agent' => 'MTG-Estimator/1.0')
  end

  # Search for a card by name using Scryfall API
  # @param card_name [String] Name of the card to search for
  # @return [Hash, nil] Card data from Scryfall or nil if not found
  def search_card_by_name(card_name)
    response = self.class.get('/cards/named', query: { fuzzy: card_name })
    
    return response.parsed_response if response.success?
    
    nil
  rescue StandardError => e
    puts "Error searching for card: #{e.message}"
    nil
  end

  # Get detailed card information by Scryfall ID
  # @param scryfall_id [String] Scryfall UUID for the card
  # @return [Hash, nil] Card data from Scryfall or nil if not found
  def get_card_info(scryfall_id)
    response = self.class.get("/cards/#{scryfall_id}")
    
    return response.parsed_response if response.success?
    
    nil
  rescue StandardError => e
    puts "Error getting card info: #{e.message}"
    nil
  end

  # Search for cards using Scryfall search syntax
  # @param query [String] Search query
  # @param page [Integer] Page number for pagination
  # @return [Hash] Search results from Scryfall
  def search_cards(query, page = 1)
    response = self.class.get('/cards/search', query: { q: query, page: page })
    
    return response.parsed_response if response.success?
    
    { 'object' => 'error', 'details' => 'Search failed' }
  rescue StandardError => e
    puts "Error searching cards: #{e.message}"
    { 'object' => 'error', 'details' => e.message }
  end
end
